parsley-debug/shared/src/main/scala/parsley/debugger/internal/TransientDebugTree.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2020 Parsley Contributors <https://github.com/j-mie6/Parsley/graphs/contributors>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+package parsley.debugger.internal
+
+import scala.collection.immutable.ListMap
+import scala.collection.mutable
+
+import parsley.debugger.{DebugTree, ParseAttempt}
+
+/** A mutable implementation of [[DebugTree]], used when constructing the tree as a parser is
+  * running.
+  *
+  * When viewing / analysing the parse tree, it is highly advised to call
+  * [[TransientDebugTree#freeze]] to obtain a frozen, immutable version of the debug tree.
+  *
+  * @param name Name of parser.
+  * @param parse What attempts to parse have been made?
+  * @param children This debug tree node's children.
+  */
+private [parsley] case class TransientDebugTree(
+    var name: String = "",
+    var internal: String = "",
+    fullInput: String,
+    var parse: Option[ParseAttempt] = None,
+    var cNumber: Option[Long] = None,
+    children: mutable.Map[String, TransientDebugTree] = new mutable.LinkedHashMap()
+) extends DebugTree {
+    // These are user-facing, and will depend heavily on what the parser looks like.
+    // $COVERAGE-OFF$
+    override def parserName: String = name
+
+    override def internalName: String = internal
+
+    override def childNumber: Option[Long] = cNumber
+
+    // The pair stores the input the parser attempted to parse and its success.
+    override def parseResults: Option[ParseAttempt] = parse
+
+    // Provides a consistent view on the children of the tree.
+    // There shouldn't really be a need to call into the expensive path often, but this is guarding against
+    // concurrent accesses (if the platform supports synchronized).
+    private var lastChildrenView: Map[String, DebugTree] = Map.empty
+    private def childrenOrGenerate(): Map[String, DebugTree] = this.synchronized {
+        if (lastChildrenView.size != children.size) {
+            lastChildrenView = children.foldLeft[ListMap[String, DebugTree]](new ListMap())(_ + _)
```

Comment:
`children.foldLeft(ListMap.empty[String, DebugTree])(_ + _)`

Reply:
Ideally I'd replace that entire mechanism with something that returns an `immutable.Map[String, DebugTree]` that uses `children` as a backing map, but there doesn't seem to be something like that in Scala unless I inherit from `Map` (which has different abstract methods between 2.12 and 2.13).

Reply:
Use children for structural sharing you mean?

Reply:
See: https://github.com/j-mie6/parsley/blob/d53d07bd45ecaadb0ce268e7af90924453f15737/parsley-debug/shared/src/main/scala/parsley/debugger/internal/TransientDebugTree.scala#L37

Reply:
For now, I've re-added `XMap`, since that seemed to be a lightweight fix for now (also made it `private [parsley]`, as it doesn't need to be public API.